container_crawler/__init__.py

```
@@ -86,40 +139,73 @@ def _worker(self):
+                    row['ROWID'], row['name'].decode('utf-8'),
+                    traceback.format_exc()))
             finally:
                 self.work_queue.task_done()
 
-    def _stop(self):
-        for _ in xrange(self.workers):
-            self.work_queue.put(None)
-        self.pool.waitall()
-
-    def _check_errors(self):
-        # When working in bulk, errors are propagated immediately
-        if self.bulk or self.error_queue.empty():
-            return
+    def _enumerator(self):
+        job = ContainerJob()
+        while 1:
+            try:
+                work = self.enumerator_queue.get()
+            except:
+                self.log(
+                    'error', 'Failed to fetch containers to enumerate %s' %
+                    traceback.format_exc())
+                time.sleep(100)
+                continue
 
-        retry_error = False
+            try:
+                if not work:
+                    break
+
+                settings, per_account = work
+                handler = self.handler_class(self.status_dir, settings,
+                                             per_account=per_account)
+                owned, verified, last_row, db_id = self.handle_container(
+                    handler, job)
+                if not owned and not verified:
+                    continue
 
-        while not self.error_queue.empty():
-            row, error = self.error_queue.get()
-            if error:
-                self.log('error', u'Failed to handle row %s (%s): %r' % (
-                    row['ROWID'], row['name'].decode('utf-8'), error))
-            else:
-                retry_error = True
-        if not retry_error:
-            raise RuntimeError('Failed to process rows')
-        else:
-            raise RetryError('Rows must be retried later')
+                if self.bulk or job.wait_all() == ContainerJob.PASS_SUCCEEDED:
+                    handler.save_last_row(last_row, db_id)
```

Wondering what the contract is here. If there is a failure on the save_last_row, does it throw an exception or fail quietly? Is the traceback in the log sufficient for this case?

---

`save_last_row` is defined by the handler class, so it could do anything. In our case, though, with 1space, it does raise an exception which will be caught by the catch-all `except` clause on line 198 and we get back `traceback.format_exc()`, which should include the full stack trace.

---

If that's the intent, then this package should provide an exception for the user of this package to raise, and trap that here.  1space and whoever else can use it.  In any case, this would presumably be a subclass of `Exception`, anyway, not requiring a bare except.

Then, unexpected exceptions can be handled as appropriate... which is usually letting non-Exception-inheriting exceptions get past and causing some kind of process exit.  The concern, here, would be continuing to try to operate when the system is no longer in a known state, because some unexpected exception came out of `save_last_row`.

---

(oops, i assumed this was a reply to my bare exception question--please take w/a grain of salt)